languages/ruby/exercises/concept/nil/.meta/example.rb

```diff
@@ -0,0 +1,8 @@
+class NameBadge
+  def self.print(id, name, department)
+    department ||= 'owner'
+
+    prefix = if id.nil? then '' else "[#{id}] - " end
```

Comment:
There are bugs in this name badge and it is specifically because of line 5.  Should we come up with a different example?  The form is incorrect, it allows a result of `"[false]"`.  And we would probably interpolate `nil` as an empty string, which could be interpolated to create a consistent badge design of `"[]"` with the brackets being empty, if nil, or populated by a different id (even false) if not nil.

If false is invalid and nil is invalid, we would probably do something like:

```ruby
prefix = if id then "[#{id}]" else ''
```
If prefix could be nil, then we might do something like:

```ruby
prefix = "[#{id}]" unless id.nil?
```

The use here seems unidiomatic, otherwise, and perhaps buggy.

Reply:
I don't agree. We shouldn't need to take care of all values passed in. We're also not taking care of `<SomeClass#objectid>`, nor shouldn't we.

Reply:
> We're also not taking care of <SomeClass#objectid>, nor shouldn't we.

We are taking care of `AnyObject#any_thing` though.  Perhaps we shouldn't.  But that isn't the focus here. It just works, If you give it false, the badge looks like:

```
[false] - Person - R&D
```

If you were to give it `NameBadge.new` an example of something that does not have a "nice" `to_s` method defined:

```ruby
puts NameBadge.print(NameBadge.new, 'Person', 'R&D')
```
Gives me the output of:

```
[#<NameBadge:0x0000559f21bd8b50>] - Person - R&D
```

Reply:
So, I'm not exactly sure where to go from here.  As I read the issue/design, this exercise is to depend only on concepts introduced in basics (class definition, method definition, arithmetic, returning a value) and booleans (true, false, logical operators, ternary expression).  I think you could even say that it shouldn't even use strings, string literals, or string interpolation (so may indeed require a different example unless the design is changed to include that as a pre-requisite).

Though, I'm not sure the scope of any of the language concept exercises at these very early points are able to handle the defensive programming approach critique, which I think (correct me if wrong) is what your guidance is aimed at, @kotp. 

So I guess what I am going to ask is a few things to suss out an approach to this exercise's coding problem:
1) Where in ruby to do often encounter `nil` values in your experience?
   - Are there certain classes or objects that you find them used more often?
   - In what circumstances do you find `nil` often being returned from methods? 
2) Where do you find yourself reaching for `#nil?` to get a boolean value?
3) If not a string application, because it may be prone to errors due to unintended parameters passed to it, is there another domain which could be more applicable?

> I am totally willing to do the work, research, etc, but being on the new side to ruby I could definitely use some help getting some keywords/direction for getting this exercise to a better place to teach idiomatic ruby :smile: